ocrd_utils/ocrd_utils/__init__.py

```diff
@@ -267,17 +286,53 @@ def crop_image(image, box=None):
 
     Return a new PIL.Image.
     """
-    # todo: perhaps we should issue a warning if we encounter this
-    # (It should be invalid in PAGE-XML to extend beyond parents.)
     if not box:
         box = (0, 0, image.width, image.height)
+    elif box[0] < 0 or box[1] < 0 or box[2] > image.width or box[3] > image.height:
+        # (It should be invalid in PAGE-XML to extend beyond parents.)
+        LOG.warning('crop coordinates (%s) exceed image (%dx%d)',
+                    str(box), image.width, image.height)
     xywh = xywh_from_bbox(*box)
     background = ImageStat.Stat(image).median[0]
     new_image = Image.new(image.mode, (xywh['w'], xywh['h']),
                           background) # or 'white'
     new_image.paste(image, (-xywh['x'], -xywh['y']))
     return new_image
 
+def rotate_image(image, angle, fill='background', transparency=False):
+    """"Rotate an image, enlarging and filling with background.
+
+    Given a PIL.Image ``image`` and a rotation angle in degrees
+    counter-clockwise ``angle``, rotate the image, increasing its
+    size at the margins accordingly, and filling everything outside
+    the original image according to ``fill``:
+    - if ``background`` (the default),
+      then use the median color of the image;
+    - otherwise use the given color, e.g. ``white`` or (255,255,255).
+    Moreover, if ``transparency`` is true, then add an alpha channel
+    fully opaque (i.e. everything outside the original image will
+    be transparent for those that can interpret alpha channels).
+    (This is true for images which already have an alpha channel,
+    regardless of the setting used.)
+
+    Return a new PIL.Image.
+    """
+    if fill == 'background':
+        background = ImageStat.Stat(image).median[0]
+    else:
+        background = fill
```

Comment:
Will it really work with strings like ``white``? Not likely. So, we have a parameter which can be a string (`default`) or a tuple (or a list)?

Reply:
> Will it really work with strings like `white`?

Yes, that does work (in all colour modes). If a tuple is used, its dimensionality must match the colour mode (e.g. 4 for RGBA).

It does not hurt to pass PIL's colour specifier through here, and I found it useful for debugging purposes (using 'red'). But it might be actually useful for `P` images (colour palette) with a "special colour".

Reply:
:+1: